pkg/kubectl/resource/interfaces.go

```diff
@@ -44,3 +44,41 @@ type ClientMapperFunc func(mapping *meta.RESTMapping) (RESTClient, error)
 func (f ClientMapperFunc) ClientForMapping(mapping *meta.RESTMapping) (RESTClient, error) {
 	return f(mapping)
 }
+
+// RequestOptions is a function that is given a chance to modify the outgoing request.
+type RequestOptions func(*client.Request)
+
+// NewClientWithOptions wraps the provided RESTClient and invokes each opt on each
+// newly created request.
+func NewClientWithOptions(c RESTClient, opts ...RequestOptions) RESTClient {
+	return &clientOptions{c: c, opts: opts}
+}
+
+type clientOptions struct {
+	c    RESTClient
+	opts []RequestOptions
+}
+
+func (c *clientOptions) modify(req *client.Request) *client.Request {
+	for _, opt := range c.opts {
+		opt(req)
+	}
+	return req
+}
+
+func (c *clientOptions) Get() *client.Request {
+	return c.modify(c.c.Get())
+}
+
+func (c *clientOptions) Post() *client.Request {
```

Comment:
should we have separate transformation fns for each operation type?

Reply:
Would wait for a use case for it.  As I noted in the commit, we may want to know what the request looks like for applying, but don't have a concrete use case.  For now, the two use cases are:

1. server-side get - request a specific content type via accept header
2. client builder overriding throttle to do admin operations

Easy enough to change later.

Reply:
ACK